### PR TITLE
Finish up JsonWriter Kotlin conversion

### PR DESCRIPTION
**Pull request by ZacSweers**
Finishing up #1477 to proceed with other work

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.kt [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    when (value) {
+      is Map<*, *> -> {
+        beginObject()
+        for ((key, value1) in value) {
+          require(key is String) {
+            if (key == null) {
+              "Map keys must be non-null"
+            } else {
+              "Map keys must be of type String: ${key.javaClass.name}"
+            }
```
Comment (JvmName) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.kt:401]: nit -- i feel like this'd be more idiomatic as 2 `require`s? 
```suggestion
		 requireNotNull(key) { "Map keys must be not-null" } 
		 require(key is String) { "Map keys must be of type String: ${key.javaClass.name}" }
```

**Review by JvmName — APPROVED** [review]

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.kt [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+   *
+   * @return this writer.
+   * @see JsonReader.readJsonValue
+   */
+  @Throws(IOException::class)
+  public fun jsonValue(value: Any?): JsonWriter {
+    when (value) {
+      is Map<*, *> -> {
+        beginObject()
+        for ((key, value1) in value) {
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.kt]: e0477a4

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.kt (hunk 1 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.kt]: dig, though github won't let me add suggestion from here. Let me push

Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.kt]: 16de888

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.kt [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+  protected var pathNames: Array<String?> = arrayOfNulls(32)
+
+  @JvmField
+  protected var pathIndices: IntArray = IntArray(32)
+
+  /**
+   * A string containing a full set of spaces for a single level of indentation, or null for no
+   * pretty printing.
+   */
+  @JvmField
```
Comment (Goooler) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.kt:153]: Seems these `@JvmField` could be removed now?

**Pull request merged**